"""
QQBot gRPCæœåŠ¡å™¨å®ç°
"""
import json
import time
import grpc
from concurrent import futures
from dataclasses import asdict
from loguru import logger

import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from src.qqbot.sender import QQBotSender
from src.qqbot import qqbot_pb2, qqbot_pb2_grpc


class QQBotService(qqbot_pb2_grpc.QQBotServiceServicer):
    """
    QQBot gRPCæœåŠ¡å®ç°
    """

    def __init__(self, sender: QQBotSender):
        self.sender = sender

    async def SendMessage(self, request, context):
        """å‘é€å•æ¡æ¶ˆæ¯"""
        logger.info(f"ğŸ“¨ å‘é€QQæ¶ˆæ¯è¯·æ±‚: {request.target_type} {request.target_id}")

        try:
            # è§£ææ¶ˆæ¯å†…å®¹
            content = None

            if request.content.HasField("text"):
                content = request.content.text
            elif request.content.HasField("segment"):
                # å•ä¸ªæ¶ˆæ¯æ®µ
                segment_data = {
                    "type": request.content.segment.type,
                    "data": json.loads(request.content.segment.data) if request.content.segment.data else {}
                }
                content = segment_data
            elif request.content.segments:
                # å¤šä¸ªæ¶ˆæ¯æ®µ
                segments_data = []
                for segment in request.content.segments:
                    segment_dict = {
                        "type": segment.type,
                        "data": json.loads(segment.data) if segment.data else {}
                    }
                    segments_data.append(segment_dict)

                content = {
                    "type": "MessageArray",
                    "data": segments_data
                }

            if not content:
                return qqbot_pb2.SendMessageResponse(
                    status=400,
                    message="æ¶ˆæ¯å†…å®¹ä¸èƒ½ä¸ºç©º",
                    data=json.dumps({"error": "Empty content"}, ensure_ascii=False)
                )

            # æ„å»ºå…ƒæ•°æ®
            metadata = dict(request.metadata)
            if request.sender_id:
                metadata['sender_id'] = request.sender_id

            # å‘é€æ¶ˆæ¯
            result = await self.sender.send_message(
                target_id=request.target_id,
                target_type=request.target_type,
                content=content,
                metadata=metadata
            )

            # ä½¿ç”¨dataclassçš„asdictæ–¹æ³•
            result_dict = asdict(result)

            # æ„å»ºå“åº”
            status_code = 200 if result.success else 500

            return qqbot_pb2.SendMessageResponse(
                status=status_code,
                message=result.message,
                data=json.dumps(result_dict, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¤„ç†å‘é€æ¶ˆæ¯è¯·æ±‚å¤±è´¥: {e}")

            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "target_id": request.target_id,
                "target_type": request.target_type,
                "success": False
            }

            return qqbot_pb2.SendMessageResponse(
                status=500,
                message=f"å†…éƒ¨æœåŠ¡å™¨é”™è¯¯: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def SendBatchMessages(self, request, context):
        """æ‰¹é‡å‘é€æ¶ˆæ¯"""
        logger.info(f"ğŸ“¦ æ‰¹é‡å‘é€QQæ¶ˆæ¯è¯·æ±‚ï¼Œæ•°é‡: {len(request.target_ids)}")

        try:
            # è§£ææ¶ˆæ¯å†…å®¹
            content = None

            if request.content.HasField("text"):
                content = request.content.text
            elif request.content.HasField("segment"):
                # å•ä¸ªæ¶ˆæ¯æ®µ
                segment_data = {
                    "type": request.content.segment.type,
                    "data": json.loads(request.content.segment.data) if request.content.segment.data else {}
                }
                content = segment_data
            elif request.content.segments:
                # å¤šä¸ªæ¶ˆæ¯æ®µ
                segments_data = []
                for segment in request.content.segments:
                    segment_dict = {
                        "type": segment.type,
                        "data": json.loads(segment.data) if segment.data else {}
                    }
                    segments_data.append(segment_dict)

                content = {
                    "type": "MessageArray",
                    "data": segments_data
                }

            if not content:
                return qqbot_pb2.SendBatchMessagesResponse(
                    status=400,
                    message="æ¶ˆæ¯å†…å®¹ä¸èƒ½ä¸ºç©º",
                    data=json.dumps({"error": "Empty content"}, ensure_ascii=False)
                )

            # æ„å»ºå…ƒæ•°æ®
            metadata = dict(request.metadata)
            if request.sender_id:
                metadata['sender_id'] = request.sender_id

            # æ‰¹é‡å‘é€æ¶ˆæ¯
            result = await self.sender.send_batch_messages(
                target_ids=list(request.target_ids),
                target_type=request.target_type,
                content=content,
                metadata=metadata
            )

            # æ„å»ºå“åº”
            overall_success = result.get('success_count', 0) > 0
            status_code = 200 if overall_success else 500

            return qqbot_pb2.SendBatchMessagesResponse(
                status=status_code,
                message=f"æ‰¹é‡å‘é€å®Œæˆï¼ŒæˆåŠŸ {result.get('success_count', 0)} æ¡ï¼Œå¤±è´¥ {result.get('failure_count', 0)} æ¡",
                data=json.dumps(result, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¤„ç†æ‰¹é‡å‘é€è¯·æ±‚å¤±è´¥: {e}")

            error_data = {
                "error": str(e),
                "timestamp": time.time(),
                "target_ids_count": len(request.target_ids),
                "target_type": request.target_type,
                "success": False
            }

            return qqbot_pb2.SendBatchMessagesResponse(
                status=500,
                message=f"å†…éƒ¨æœåŠ¡å™¨é”™è¯¯: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )

    async def HealthCheck(self, request, context):
        """å¥åº·æ£€æŸ¥"""
        try:
            # æ£€æŸ¥QQBotçŠ¶æ€
            health_status = await self.sender.health_check()

            # ä½¿ç”¨dataclassçš„asdictæ–¹æ³•
            health_dict = asdict(health_status)

            health_data = {
                "timestamp": time.time(),
                "service_ready": health_status.healthy,  # ç›´æ¥è®¿é—®å±æ€§
                "health_status": "healthy" if health_status.healthy else "unhealthy",
                "details": health_dict
            }

            status_code = 200 if health_status.healthy else 503

            return qqbot_pb2.HealthCheckResponse(
                status=status_code,
                message="æœåŠ¡å¥åº·" if health_status.healthy else "æœåŠ¡ä¸å¥åº·",
                data=json.dumps(health_data, ensure_ascii=False)
            )

        except Exception as e:
            logger.error(f"ğŸ’¥ å¥åº·æ£€æŸ¥å¤±è´¥: {e}")

            error_data = {
                "timestamp": time.time(),
                "service_ready": False,
                "error": str(e),
                "details": "å¥åº·æ£€æŸ¥å¼‚å¸¸"
            }

            return qqbot_pb2.HealthCheckResponse(
                status=500,
                message=f"å¥åº·æ£€æŸ¥å¤±è´¥: {str(e)}",
                data=json.dumps(error_data, ensure_ascii=False)
            )


def create_server(sender: QQBotSender, max_workers: int = 10) -> grpc.aio.Server:
    """
    åˆ›å»ºgRPCæœåŠ¡å™¨

    Args:
        sender: QQæ¶ˆæ¯å‘é€å™¨
        max_workers: æœ€å¤§å·¥ä½œçº¿ç¨‹æ•°

    Returns:
        gRPCæœåŠ¡å™¨å®ä¾‹
    """
    server = grpc.aio.server(
        futures.ThreadPoolExecutor(max_workers=max_workers)
    )

    # æ·»åŠ æœåŠ¡
    qqbot_service = QQBotService(sender)
    qqbot_pb2_grpc.add_QQBotServiceServicer_to_server(qqbot_service, server)

    return server
